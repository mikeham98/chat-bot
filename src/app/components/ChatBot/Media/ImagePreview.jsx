import React from 'react';
import styles from '../../../../themes/components/media.scss';

const ImagePreview = ({src}) => {
    return (
        <img className={styles.mediaImagePreviewImage} src={src}/>
    );
};

export default React.memo(ImagePreview);