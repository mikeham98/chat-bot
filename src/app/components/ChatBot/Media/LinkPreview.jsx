import React from 'react';
import styles from '../../../../themes/components/media.scss';
import PropTypes from "prop-types";

const LinkPreview = ({src, image, title, description}) => {
    return (
        <div className={styles.mediaLinkPreviewWrapper} onClick={() => window.open(src)}>
            <img src={image}/>
            <div className={styles.mediaLinkPreviewMeta}>
                <span className={styles.mediaLinkPreviewTitle}>{title}</span>
                <span className={styles.mediaLinkPreviewDescription}>{description}</span>
            </div>
        </div>
    );
};

LinkPreview.propTypes = {
    src: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
};

export default React.memo(LinkPreview);