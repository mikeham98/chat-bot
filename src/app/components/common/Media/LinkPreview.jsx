import React from 'react';
import styles from '../../../../themes/components/media.scss';

const LinkPreview = ({src, image, title, description}) => {
    return (
        <div className={styles.mediaLinkPreviewWrapper} onClick={() => window.open(src)}>
            <img src={image}/>
            <div className={styles.mediaLinkPreviewMeta}>
                <span className={styles.mediaLinkPreviewTitle}>{title}</span>
                <span className={styles.mediaLinkPreviewDescription}>{description}</span>
            </div>
        </div>
    );
};

export default React.memo(LinkPreview);