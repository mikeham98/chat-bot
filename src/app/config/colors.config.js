const flatRed = '#f4244e';
const flatOrange = '#f87b2b';
const flatYellow = '#F1C40E';
const flatGreen = '#30CB71';
const flatTeal = '#1ccb9e';
const flatBlue = '#3598DA';
const flatLavender = '#6872FF';
const flatPurple = '#9A58B9';

export const flatColours = [
    flatRed,
    flatOrange,
    flatYellow,
    flatGreen,
    flatTeal,
    flatBlue,
    flatLavender,
    flatPurple
];