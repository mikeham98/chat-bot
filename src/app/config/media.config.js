export const gif = 'gif';
export const img = 'img';
export const link = 'link';