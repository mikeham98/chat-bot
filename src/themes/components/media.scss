@import '../colors';

.mediaLinkPreviewWrapper, .mediaImagePreviewImage {
  margin: 10px 0 5px 0;
  border-width: 1px;
  border-style: solid;
  border-color: $light-gray;
  border-radius: 1em;
  overflow: hidden;
}

.mediaLinkPreviewWrapper > img, .mediaImagePreviewImage {
  object-fit: cover;
}

.mediaImagePreviewImage {
  margin: 10px 0 5px 0;
  height: 150px;
  width: 200px;
  overflow: hidden;
  border-radius: 1em;
}

.mediaLinkPreviewWrapper {
  cursor: pointer;
  display: flex;
  flex-direction: column;
  img {
    height: 150px;
  }
  width: 300px;
  .mediaLinkPreviewMeta {
    padding: 10px;
    background-color: white;
    flex-direction: column;
    display: flex;
    justify-content: space-evenly;
    .mediaLinkPreviewTitle {
      margin: 5px 0;
      color: $flat-lavender;
      font-weight: 600;
    }
    .mediaLinkPreviewDescription {
      color: $gray;
      font-size: 0.8em;
    }
  }
}

@media only screen and (max-width: 600px) {
  .mediaLinkPreviewMeta {
    max-width: 100%;
  }
}