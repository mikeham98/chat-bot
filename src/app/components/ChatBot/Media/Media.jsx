import React from 'react';
import ImageMedia from "./ImagePreview";
import LinkMedia from "./LinkPreview";
import isPopulatedArray from "../../../util/isPopulatedArray";
import {gif, img, link} from "../../../config/media.config";

export default class Media extends React.PureComponent {
    returnMedia() {
        const {media} = this.props;
        if (isPopulatedArray(media)) {
            return media.map(({id, type, src, imagePreview, title, description}) => {
                if (type === gif || type === img) {
                    return (
                        <ImageMedia
                            key={id}
                            src={src}
                        />
                    );
                }
                if (type === link) {
                    return (
                        <LinkMedia
                            key={id}
                            src={src}
                            image={imagePreview}
                            title={title}
                            description={description}
                        />
                    );
                }
                return null;
            });
        }
    }

    render() {
        return (
            <div>
                {this.returnMedia()}
            </div>
        );
    }
}