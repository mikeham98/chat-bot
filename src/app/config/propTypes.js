import PropTypes from 'prop-types';
import {gif, img, link} from "./media.config";

export const OptionPropTypes = PropTypes.shape({
    id: PropTypes.number.isRequired,
    option: PropTypes.string.isRequired,
});

export const MediaObjectPropTypes = PropTypes.shape({
    id: PropTypes.number.isRequired,
    src: PropTypes.string.isRequired,
    type: PropTypes.oneOf([gif, img, link]).isRequired,
    imagePreview: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string,
});

export const MediaPropTypes = PropTypes.arrayOf(MediaPropTypes);

export const OptionsPropTypes = PropTypes.arrayOf(OptionPropTypes);

export const MessagesPropTypes = PropTypes.arrayOf(PropTypes.shape({
    content: PropTypes.shape({
        body: PropTypes.string.isRequired,
        media: MediaPropTypes,
        options: OptionsPropTypes,
    }),
    recipient: PropTypes.shape({
        userId: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        userName: PropTypes.string
    }),
    createdAt: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.instanceOf(Date)
    ]).isRequired,
    createdBy: PropTypes.shape({
        userId: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        userName: PropTypes.string,
    }).isRequired
}));