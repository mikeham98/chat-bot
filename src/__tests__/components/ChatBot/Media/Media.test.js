import React from 'react';
import {shallow} from 'enzyme';
import Media from '../../../../app/components/ChatBot/Media/Media';
import {gif, img, link} from "../../../../app/config/media.config";

describe('Media', () => {

    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = shallow(<Media {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            media
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return two ImagePreview components (gif and img) and a LinkPreview component', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return 3 media components even though a 4th has been added but with a unknown type', () => {
            const nullMedia = {
                id: 4,
                type: 'random',
                src: 'https://images.unsplash.com/image.jpg'
            };
            props.media = [...media, nullMedia];
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return an empty div when there is no media', () => {
            props.media = [];
            expect(wrapper()).toMatchSnapshot();
        });
    });
});

const media = [{
    id: 1,
    type: link,
    imagePreview: 'https://images8.alphacoders.com/374/374852.jpg',
    title: 'How (and Why) SpaceX Will Colonize Mars',
    description: 'A brief insight to how and why SpaceX will Colonize Mars',
    src: 'https://waitbutwhy.com/2015/08/how-and-why-spacex-will-colonize-mars.html'
}, {
    id: 2,
    type: gif,
    src: 'https://media.giphy.com/media/1108D2tVaUN3eo/giphy.gif'
}, {
    id: 3,
    type: img,
    src: 'https://images.unsplash.com/photo-1564053489984-317bbd824340?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1028&q=80'
}];