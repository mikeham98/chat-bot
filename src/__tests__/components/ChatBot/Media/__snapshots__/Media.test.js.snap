// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media render should return an empty div when there is no media 1`] = `<div />`;

exports[`Media render should return two ImagePreview components (gif and img) and a LinkPreview component 1`] = `
<div>
  <Memo(LinkPreview)
    description="A brief insight to how and why SpaceX will Colonize Mars"
    image="https://images8.alphacoders.com/374/374852.jpg"
    key="1"
    src="https://waitbutwhy.com/2015/08/how-and-why-spacex-will-colonize-mars.html"
    title="How (and Why) SpaceX Will Colonize Mars"
  />
  <Memo(ImagePreview)
    key="2"
    src="https://media.giphy.com/media/1108D2tVaUN3eo/giphy.gif"
  />
  <Memo(ImagePreview)
    key="3"
    src="https://images.unsplash.com/photo-1564053489984-317bbd824340?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1028&q=80"
  />
</div>
`;
